src/main/java/com/appdynamics/monitors/azure/metrics/AzureMetrics.java

```diff
@@ -0,0 +1,185 @@
+ *  The copyright notice above does not evidence any actual or intended publication of such source code.
+ *
+ */
+
+package com.appdynamics.monitors.azure.metrics;
+
+import com.appdynamics.extensions.AMonitorTaskRunnable;
+import com.appdynamics.extensions.MetricWriteHelper;
+import com.appdynamics.extensions.metrics.Metric;
+import com.appdynamics.monitors.azure.utils.Constants;
+import com.google.common.collect.Lists;
+import com.microsoft.azure.management.Azure;
+import com.microsoft.azure.management.monitor.MetricCollection;
+import com.microsoft.azure.management.monitor.MetricDefinition;
+import com.microsoft.azure.management.resources.GenericResource;
+import org.joda.time.DateTime;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+
+@SuppressWarnings("unchecked")
+public class AzureMetrics implements AMonitorTaskRunnable {
+    private static final org.slf4j.Logger logger = LoggerFactory.getLogger(AzureMetrics.class);
+    private final Map<String, ?> resourceFilter;
+    private final String currentResourceGroupFilter;
+    private final String currentResourceFilter;
+    private final String currentResourceTypeFilter;
+    private final GenericResource resource;
+    private final String subscriptionName;
+    private final CountDownLatch countDownLatch;
+    private final MetricWriteHelper metricWriteHelper;
+    private final Azure azure;
+    private final String metricPrefix;
+
+    public AzureMetrics(Map<String, ?> resourceFilter,
+                        String currentResourceGroupFilter,
+                        String currentResourceFilter,
+                        String currentResourceTypeFilter,
+                        GenericResource resource,
+                        String subscriptionName,
+                        CountDownLatch countDownLatch,
+                        MetricWriteHelper metricWriteHelper,
+                        Azure azure,
+                        String metricPrefix) {
+        this.resourceFilter = resourceFilter;
+        this.currentResourceGroupFilter = currentResourceGroupFilter;
+        this.currentResourceFilter = currentResourceFilter;
+        this.currentResourceTypeFilter = currentResourceTypeFilter;
+        this.resource = resource;
+        this.subscriptionName = subscriptionName;
+        this.countDownLatch = countDownLatch;
+        this.metricWriteHelper = metricWriteHelper;
+        this.azure = azure;
+        this.metricPrefix = metricPrefix;
+    }
+
+    @Override
+    public void onTaskComplete() {
+        logger.info("Task Complete");
+    }
+
+    @Override
+    public void run() {
+        try {
+            runTask();
+        }
+        catch(Exception e){
+            logger.error(e.getMessage());
+        }
+        finally {
+            countDownLatch.countDown();
+        }
+    }
+    private void runTask(){
+        List<Metric> finalMetricList = Lists.newArrayList();
```

Comment:
I highly recommend breaking this method down into a couple of other methods. This would make the code more modular and a lot easier to write unit tests against.